crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -0,0 +1,117 @@
+use {
+    super::{NativePriceEstimateResult, NativePriceEstimating},
+    crate::price_estimation::PriceEstimationError,
+    anyhow::{anyhow, Result},
+    futures::{future::BoxFuture, FutureExt},
+    primitive_types::H160,
+    reqwest::{header::AUTHORIZATION, Client, StatusCode},
+    serde::Deserialize,
+    std::collections::HashMap,
+    url::Url,
+};
+
+#[derive(Debug, Deserialize)]
+struct Response(HashMap<H160, Price>);
+
+#[derive(Debug, Deserialize)]
+struct Price {
+    eth: f64,
+}
+
+type Token = H160;
+
+pub struct CoinGecko {
+    client: Client,
+    base_url: Url,
+    api_key: Option<String>,
+    chain: String,
+}
+
+impl CoinGecko {
+    pub fn new(
+        client: Client,
+        base_url: Url,
+        api_key: Option<String>,
+        chain_id: u64,
+    ) -> Result<Self> {
+        let chain = match chain_id {
+            1 => "ethereum".to_string(),
+            100 => "xdai".to_string(),
+            42161 => "arbitrum-one".to_string(),
+            n => anyhow::bail!("unknown network {n}"),
+        };
+        Ok(Self {
+            client,
+            base_url,
+            api_key,
+            chain,
+        })
+    }
+}
+
+impl NativePriceEstimating for CoinGecko {
+    fn estimate_native_price(&self, token: Token) -> BoxFuture<'_, NativePriceEstimateResult> {
+        async move {
+            let mut builder = self.client.get(format!(
+                "{}/simple/token_price/{}?contract_addresses=0x{token:x}&vs_currencies=eth",
+                self.base_url, self.chain
+            ));
+            if let Some(ref api_key) = self.api_key {
+                builder = builder.header(AUTHORIZATION, api_key)
+            }
+            let response = builder.send().await.map_err(|e| {
+                PriceEstimationError::EstimatorInternal(anyhow!(
+                    "failed to sent CoinGecko price request: {e:?}"
+                ))
+            })?;
+            if !response.status().is_success() {
+                let status = response.status();
+                tracing::warn!(?status, "CoinGecko price request failed",);
+                return match status {
+                    StatusCode::TOO_MANY_REQUESTS => Err(PriceEstimationError::RateLimited),
+                    _ => Err(PriceEstimationError::NoLiquidity),
+                };
+            }
+            let response = response.text().await.map_err(|e| {
+                PriceEstimationError::EstimatorInternal(anyhow!(
+                    "failed to fetch Native CoinGecko prices: {e:?}"
+                ))
+            })?;
+            let prices = serde_json::from_str::<Response>(&response)
```

Comment:
We do not need to normalize prices since CoinGecko already returns the prices of 1 unit of the quotes amount, so if we request how much `eth` is `weth`, its response is `1.0` (or around it, since it is not exactly `1.0`), is that exactly what we want, right?